Easy to use gitian building script

achow101 opened the pull request:
I've written a script that automates the setup and building of binaries with gitian. All of the commands are pulled from various documentation on gitian building.

The script allows you to specify unsigned or signed builds, which os to build for, the signer, the version (or commit/branch), and build arguments (threads and memory).

Right now setup is designed for debian based systems.

It runs a few commands that require sudo. For best automation, add apt-get, lxc-start, lxc-execute, and ifconfig to your sudoers file.

achow101 committed:
Create an easy to use gitian building script

I've written a script that automates the setup and building of binaries with gitian. All of the commands are pulled from various documentation on gitian building.

jonasschnelli commented:
Nice!
Concept ACK.

isle2983 commented:
Comment spelling/grammar nits:

s/OtherBasic/Other basic/ 
s/Debian based/Debian-based/
s/arguement/argument/
s/# Window/# Windows/
s/# Sign Sindows/# Sign Windows/

achow101 commented:
Nits addressed

laanwj commented:
Concept ACK, haven't tested

btcdrak commented:
Concept ACK, I'm testing this out.

sipa commented:
One feature I've always wanted (but never bothered to add to my own build scripts) is detached signing: I don't want my GPG key on the build machine. However, if you don't use gsign, the assert file is overwritten for every build. So if possible I'd like to have a means to just copy the assert file to the right location in gitian.sigs, but without adding the gpg signature yet.

btcdrak commented on contrib/gitian-build.sh:
```diff
@@ -0,0 +1,350 @@
+-j		Number of processes to use. Default 4
+-m		Memory to allocate in MiB. Default 2000
+--lxc           Use LXC instead of KVM
+--setup         Setup the gitian building environment. Uses KVM. If you want to use lxc, use the --lxc option. Only works on Debian-based systems (Ubuntu, Debian)
```
This should default to lxc because it is the default according to the instructions we ship.

btcdrak commented on contrib/gitian-build.sh:
```diff
@@ -0,0 +1,350 @@
+# Check that a signer is specified
+if [[ $SIGNER == "" ]]
+then
+    echo "Must specify a signer"
```
if signer is missing you should display the usage

for example the `ln` command spits out
```
$ ln
ln: missing file operand
Try 'ln --help' for more information.
```

So I suggest:

```
$0: Missing signer
Usage: $0 [-c|u|v|b|s|B|o|h|j|m|] signer version
```

btcdrak commented on contrib/gitian-build.sh:
```diff
@@ -0,0 +1,350 @@
+# Check that a version is specified
+if [[ $VERSION == "" ]]
+then
+    echo "Must specify version"
```
Suggest

```
$0: Missing version
Usage: $0 [-c|u|v|b|s|B|o|h|j|m|] signer version
```

btcdrak commented on contrib/gitian-build.sh:
```diff
@@ -0,0 +1,350 @@
+
+# Help Message
+read -d '' usage <<- EOF
+Usage: gitian-build.sh [-c|u|v|b|s|B|o|h|j|m|] signer version
```
Consider using `$0` in place of `gitian-build.sh`

btcdrak commented on contrib/gitian-build.sh:
```diff
@@ -0,0 +1,350 @@
+	pushd ./gitian-builder	
+	mkdir -p inputs
+	wget -N -P inputs https://bitcoincore.org/cfields/osslsigncode-Backports-to-1.7.1.patch
+	wget -N -P inputs http://downloads.sourceforge.net/project/osslsigncode/osslsigncode/osslsigncode-1.7.1.tar.gz
```
Can I suggest we make these two URLs a variable defined at the beginning of the file. If this dep is updated in the future it will make it easier to spot.

btcdrak commented:
I agree with sipa, we need a "--sign-detached" option. In my case my signing keys are on a hardware device which is not available to the VM so I have to copy (scp) the gitian.sigs/* out to my main machine to detach sign. Certainly running the --sign command works, you end up with the asserts in the gitian.sigs folder but of course it generates error messages.

btcdrak commented:
This could really do with checking that the MacOS SDK file exists in the  `gitian-builder/inputs/` folder

btcdrak commented:
Tested various options, works, makes me happy.

achow101 commented:
Addressed btcdrak's concerns.

I will see what I can do regarding detached signing.

achow101 commented:
So what exactly do you guys want with detached signing?

sipa commented:
By detached signing I mean it behaves exactly as if it's signing, and the
.assert file appears inside the gitian.sigs dir, but no GPG signature for
it is created.

Thanks a lot for looking into this wishlist item :)

achow101 committed:
Check for OSX SDK

achow101 committed:
Create option to detach sign gitian builds and not commit the files in the script

Gitian builds can be done where the assert files are made but no signing is done. The sigs can also optionally not be committed automatically.

achow101 commented:
@sipa @btcdrak I think I got it to create just the assert files for detached signing. I also added an option to not automatically commit the sigs.

laanwj commented:
> By detached signing I mean it behaves exactly as if it's signing, and the
.assert file appears inside the gitian.sigs dir, but no GPG signature for
it is created.

My favorite hack for this is:
```bash
gsign -p true ...
```
(This uses the program "true" to sign, which effectively skips the step and returns success)

laanwj commented:
Oh that's what you did already, great!

laudaa commented:
"--setup" has failed because it is missing the clone bitcoin command:

> git clone https://github.com/bitcoin/bitcoin

achow101 commented:
The assumption is that you already have that


On August 26, 2016 4:03:39 PM Lauda <notifications@github.com> wrote:

> "--setup" has failed because it is missing the clone bitcoin command:
>
>> git clone https://github.com/bitcoin/bitcoin
>
> --
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/bitcoin/bitcoin/pull/8566#issuecomment-242837021

fanquake commented on contrib/gitian-build.sh:
```diff
@@ -0,0 +1,388 @@
+-B|--buildsign	Build both signed and unsigned binaries
+-o|--os		Specify which Operating Systems the build is for. Default is lwx. l for linux, w for windows, x for osx
+-j		Number of processes to use. Default 4
+-m		Memory to allocate in MiB. Default 2000
```
nit: looks like the default is actually [4000](https://github.com/bitcoin/bitcoin/pull/8566/files#diff-436c4e7496c0c2c2e1146ea5bb18079fR18).

luke-jr replied:
The default probably should be 2000. 32-bit KVM has a limit of 2047...

MarcoFalke replied:
@luke-jr this would require setting `-j` to 1 as well. I don't think those are sensible defaults for gitian building.

btcdrak replied:
It's time @luke-jr upgraded to 64 bits :8ball:

achow101 replied:
The [default from gitian](https://github.com/devrandom/gitian-builder/blob/master/bin/gbuild#L8) is `-j 2` and `-m 2000`, so I will set that to be the default.

sipa replied:
I would expect -j2 -m2000 to OOM.

achow101 replied:
It doesn't (at least, I haven't seen it do that). I have run gitian with the defaults before and it works fine, albeit rather slowly.

MarcoFalke replied:
It was causing OOM for me (on kvm, though).

luke-jr replied:
I have no problems building with -j2 -m2047 at least (KVM).

laanwj commented:
`doc/gitian-building.md` should be updated to use this script.
(not necessarily in this pull, but it makes sure that we have a "supported" build process that uses this script)

btcdrak commented:
Tested ACK 6ffd6b4a573e00644b8985c825357a7aa7b8ce88 (documentation nit aside)

achow101 committed:
Set defaults to gitian defaults

Set the default -j and -m to 2 and 2000 respectively, matching gitian's default values.
